### PR TITLE
satellite/gracefulexit: add metrics on satellite for graceful exit

### PR DESCRIPTION
**Pull request by mobyvb**
What:
`graceful_exit_init_node_age_seconds` - how old a node is when initiating graceful exit
`graceful_exit_init_node_audit_success_count` - a node's audit success count when initiating graceful exit
`graceful_exit_init_node_audit_total_count` - a node's total audit count when initiating graceful exit
`graceful_exit_init_node_piece_count` - a node's total piece count (according to `nodes` table) when initiating graceful exit
`graceful_exit_init_bytes_stored` - the initial amount of bytes placed into the transfer queue for a node after it has initiated graceful exit
`graceful_exit_transfer_piece_success` - global counter for when a transfer to a node succeeds
`graceful_exit_transfer_piece_fail` - global counter for when a transfer to a node fails
`graceful_exit_final_pieces_failed` - counter for pieces that failed to transfer during an exit
`graceful_exit_final_pieces_success` - counter for pieces that successfully transferred during an exit
`graceful_exit_final_bytes_transferred` - number of bytes successfully transferred during an exit
`graceful_exit_success` - global counter for when graceful exit completes successfully
`graceful_exit_fail_max_failures_percentage` - global counter for when a graceful exit completes but failed due to surpassing the configured "max failures percentage
`graceful_exit_fail_validation` - global counter for when a graceful exit completes but fails due to satellite validation of a piece transfer
`graceful_exit_fail_inactive` - global counter for when a graceful exit completes but fails due to storagenode inactivity
`graceful_exit_successful_pieces_transfer_ratio` - proportion of bytes successfully transferred

Why:
https://storjlabs.atlassian.net/browse/V3-2904

* reasons for how a graceful exit can complete but fail:
1. The node goes offline and is inactive for more than some configured interval (500 hours currently)
2. The node fails to transfer >10% of its pieces successfully (each piece must fail transfer 3 times before we mark it as failed)
3. The node claims to have successfully transferred a piece, but that transfer fails satellite-side validation (signatures are invalid or hashes don't match)

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/gracefulexit/endpoint.go [hunk 1]
```diff
@@ -190,14 +190,22 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
...
 			return rpcstatus.Error(rpcstatus.Internal, Error.Wrap(err).Error())
 		}
 		err = endpoint.db.IncrementProgress(ctx, nodeID, 0, 0, 0)
 		if err != nil {
 			return rpcstatus.Error(rpcstatus.Internal, Error.Wrap(err).Error())
 		}
+
+		// graceful exit initiation metrics
+		age := time.Now().UTC().Sub(node.CreatedAt.UTC())
+		mon.FloatVal("graceful_exit_init_node_age_seconds").Observe(age.Seconds())
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: I feel like with these metrics, more information is usually better. If data science wants hours or days, they can easily get that from seconds, but not the other way around.

File: satellite/gracefulexit/endpoint.go [hunk 2]
```diff
@@ -617,12 +632,15 @@ func (endpoint *Endpoint) handleSucceeded(ctx context.Context, stream processStr
 	if err != nil {
 		return Error.Wrap(err)
 	}
+
+	mon.Meter("graceful_exit_transfer_success").Mark(1)
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: Good point. Will make the name changes. I'll also ping data science and ask if these names are okay.

File: satellite/gracefulexit/endpoint.go [hunk 3]
```diff
@@ -295,7 +303,12 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 			}
 
 			var transferMsg *pb.SatelliteMessage
+			mon.IntVal("graceful_exit_final_pieces_failed").Observe(progress.PiecesFailed)
+			mon.IntVal("graceful_exit_final_pieces_succeess").Observe(progress.PiecesTransferred)
+
 			processed := progress.PiecesFailed + progress.PiecesTransferred
+			mon.IntVal("graceful_exit_successful_bytes_transfer_ratio").Observe(progress.PiecesTransferred / processed)
```
Comment (ethanadams) [inline satellite/gracefulexit/endpoint.go]: I think you want bytes transferred here, not pieces transferred

File: satellite/gracefulexit/endpoint.go [hunk 4]
```diff
@@ -348,7 +356,14 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 			}
 
 			var transferMsg *pb.SatelliteMessage
+			mon.IntVal("graceful_exit_final_pieces_failed").Observe(progress.PiecesFailed)
+			mon.IntVal("graceful_exit_final_pieces_succeess").Observe(progress.PiecesTransferred)
+
 			processed := progress.PiecesFailed + progress.PiecesTransferred
+			if processed > 0 {
+				mon.IntVal("graceful_exit_successful_bytes_transfer_ratio").Observe(progress.PiecesTransferred / processed)
```
Comment (ethanadams) [inline satellite/gracefulexit/endpoint.go]: don't you mean something like `graceful_exit_successful_pieces_transfer_ratio`

File: satellite/gracefulexit/endpoint.go (hunk 3 above)
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: It looks like we don't have a value for the total number of bytes that should have been transferred, so I don't think we can get this ratio. I changed it to `graceful_exit_successful_pieces_transfer_ratio` but I am not sure if we can get the bytes ratio with the data in the `progress` table. @benjaminsirb

File: satellite/gracefulexit/endpoint.go (hunk 4 above)
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: I changed the name.

File: satellite/gracefulexit/pathcollector.go [hunk 5]
```diff
@@ -54,15 +57,24 @@ func (collector *PathCollector) Flush(ctx context.Context) (err error) {
 
 // RemoteSegment takes a remote segment found in metainfo and creates a graceful exit transfer queue item if it doesn't exist already
 func (collector *PathCollector) RemoteSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
-	if len(collector.nodeIDs) == 0 {
+	if len(collector.nodeIDStorage) == 0 {
 		return nil
 	}
 
+	collector.nodeIDMutex.Lock()
+	defer collector.nodeIDMutex.Unlock()
```
Comment (VinozzZ) [inline satellite/gracefulexit/pathcollector.go]: nit - maybe we can unlock before the `flush` call?

Comment (VinozzZ) [reply satellite/gracefulexit/pathcollector.go]: I just noticed that the `flush` is called inside the for loop?

Comment (mobyvb) [reply satellite/gracefulexit/pathcollector.go]: Then we would have to make sure to unlock in every case that we could return inside the for loop. I am in favor of keeping the defer even if it means slightly slower processing. But I'm open to changing it if you think we should.

**Pull request merged**